src/main/java/com/avispl/symphony/dal/infrastructure/management/sembient/sembient/SembientAggregatorCommunicator.java

```diff
@@ -0,0 +1,1813 @@
+			}
+			aggregatedDevices.put(aggregatedDevice.getDeviceId(), aggregatedDevice);
+		}
+	}
+
+	/**
+	 * Populate device details information
+	 *
+	 * @param aggregatedDevice Aggregated device that get from {@link SembientAggregatorCommunicator#fetchDevicesList}
+	 */
+	void populateRegionDetails(AggregatedDevice aggregatedDevice) throws Exception {
+		// Get cached properties and controls
+		Map<String, String> properties = aggregatedDevice.getProperties();
+		List<AdvancedControllableProperty> controls = aggregatedDevice.getControllableProperties();
+		// Get current date:
+		DateTimeFormatter formatter = DateTimeFormatter.ofPattern(SembientAggregatorConstant.YYYY_MM_DD);
+		LocalDate now = LocalDate.now(ZoneId.of(SembientAggregatorConstant.UTC_TIMEZONE));
+		// Get yesterday:
+		LocalDate yesterday = LocalDate.now(ZoneId.of(SembientAggregatorConstant.UTC_TIMEZONE)).minusDays(1);
+		String currentDate = formatter.format(now);
+		String yesterdayDate = formatter.format(yesterday);
+
+		String deviceId = aggregatedDevice.getDeviceId();
+
+		String deviceName = aggregatedDevice.getDeviceName();
+		String[] rawBuildingInfo = deviceId.split(SembientAggregatorConstant.DASH);
+		int lastIndex = rawBuildingInfo.length - 1;
+		String buildingID = rawBuildingInfo[lastIndex - 2];
+		String floorName = rawBuildingInfo[lastIndex - 1];
+		if (SembientAggregatorConstant.SENSOR.equals(aggregatedDevice.getDeviceType())) {
+			// Retrieve thermal data
+			populateThermalData(properties, currentDate, yesterdayDate, buildingID, floorName, deviceName);
+			// Retrieve IAQ data
+			populateIAQData(properties, currentDate, yesterdayDate, buildingID, floorName, deviceName);
+		} else {
+			// Retrieve occupancy data
+			populateOccupancyData(properties, controls, currentDate, yesterdayDate, deviceId, buildingID, floorName, deviceName);
+			// Retrieve region tags
+			populateRegionTag(properties, controls, deviceId);
+		}
+		aggregatedDevice.setProperties(properties);
+		aggregatedDevice.setControllableProperties(controls);
+		aggregatedDevices.put(aggregatedDevice.getDeviceId(), aggregatedDevice);
+	}
+
+	/**
+	 * Populate region tag
+	 *
+	 * @param properties Map of cached properties of region (Aggregated device)
+	 * @param controls List of cached AdvancedControllableProperty of region (Aggregated device)
+	 * @param deviceId Device id of region.
+	 * @throws Exception when fail to get region tags
+	 */
+	private void populateRegionTag(Map<String, String> properties, List<AdvancedControllableProperty> controls, String deviceId) throws Exception {
+		controls.add(createText(properties, SembientAggregatorConstant.REGION_TAG_NEW_TAG, lastNewTag.get(deviceId)));
+		controls.add(createButton(properties, SembientAggregatorConstant.REGION_TAG_CREATE, SembientAggregatorConstant.LABEL_CREATE, SembientAggregatorConstant.LABEL_PRESSED_CREATING));
+		String[] regionDetails = deviceId.split(SembientAggregatorConstant.DASH);
```

Comment:
similarly to controllable properties operation - do you think the deviceId should be validated at this point, since there's a specific pattern that used for the deviceId creation

Reply:
I don't think we should validate it here. Because if we reach this method, it's the latest populated device ID.